core/src/main/java/org/apache/iceberg/io/SortedPosDeleteWriter.java

```diff
@@ -109,6 +116,8 @@ public void delete(CharSequence path, long pos, T row) {
   public List<DeleteFile> complete() throws IOException {
     close();
 
+    Preconditions.checkState(failure != null, "Cannot return results from failed writer", failure);
```

Comment:
This implements the same logic as in `BaseTaskWriter`: if there was an exception while writing the file, don't allow `complete` to succeed.